polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -107,6 +109,179 @@ struct RaiseKernelName : public OpRewritePattern<LLVM::AddressOfOp> {
                                        : std::nullopt;
   }
 };
+
+template <typename Derived, typename ConstructorOp, typename TypeTag,
+          bool PostProcess = false>
+class RaiseConstructorBasePattern : public OpRewritePattern<ConstructorOp> {
+public:
+  using OpRewritePattern<ConstructorOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ConstructorOp constructor,
+                                PatternRewriter &rewriter) const final {
+
+    if (constructor.getArgOperands().empty())
+      return failure();
+
+    // 'this*' is the first argument to the constructor call, if it is a
+    // constructor.
+    auto alloc = dyn_cast_or_null<LLVM::AllocaOp>(
+        constructor.getArgOperands().front().getDefiningOp());
+
+    if (!alloc)
+      return failure();
+
+    assert(alloc.getElemType().has_value() &&
+           "Expecting element type attribute for opaque alloca");
+
+    // Check whether the type allocated for the first argument ('this') matches
+    // the expected type.
+    auto allocTy = *alloc.getElemType();
+    auto structAllocTy = dyn_cast<LLVM::LLVMStructType>(allocTy);
+    if (!structAllocTy || structAllocTy.getName() != TypeTag::getTypeName())
+      return failure();
+
+    if (constructor.getNumResults())
+      // Constructor should not return anything.
+      return failure();
```

Comment:
Is this needed with `isConstructor` below?

Reply:
`isConstructor` would probably catch this, too. However, doing this comparably cheap check here might be more efficient to only detect this after the relatively expensive demangling.